sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PullOutGroupingExpressions.scala

```diff
@@ -0,0 +1,79 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.optimizer
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.expressions.{Alias, Expression, NamedExpression}
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateExpression
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This rule ensures that [[Aggregate]] nodes doesn't contain complex grouping expressions in the
+ * optimization phase.
+ *
+ * Complex grouping expressions are pulled out to a [[Project]] node under [[Aggregate]] and are
+ * referenced in both grouping expressions and aggregate expressions without aggregate functions.
+ * These references ensure that optimization rules don't change the aggregate expressions to invalid
+ * ones that no longer refer to any grouping expressions and also simplify the expression
+ * transformations on the node (need to transform the expression only once).
+ *
+ * For example, in the following query Spark shouldn't optimize the aggregate expression
+ * `Not(IsNull(c))` to `IsNotNull(c)` as the grouping expression is `IsNull(c)`:
+ * SELECT not(c IS NULL)
+ * FROM t
+ * GROUP BY c IS NULL
+ * Instead, the aggregate expression references a `_groupingexpression` attribute:
+ * Aggregate [_groupingexpression#233], [NOT _groupingexpression#233 AS (NOT (c IS NULL))#230]
+ * +- Project [isnull(c#219) AS _groupingexpression#233]
+ *    +- LocalRelation [c#219]
+ */
+object PullOutGroupingExpressions extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan transform {
+      case a: Aggregate if a.resolved =>
+        val complexGroupingExpressionMap = mutable.LinkedHashMap.empty[Expression, NamedExpression]
+        val newGroupingExpressions = a.groupingExpressions.map {
+          case e if !e.foldable && e.children.nonEmpty =>
+            complexGroupingExpressionMap
+              .getOrElseUpdate(e.canonicalized, Alias(e, s"_groupingexpression")())
+              .toAttribute
+          case o => o
+        }
+        if (complexGroupingExpressionMap.nonEmpty) {
+          def replaceComplexGroupingExpressions(e: Expression): Expression = {
+            e match {
+              case _ if AggregateExpression.isAggregate(e) => e
+              case _ if e.foldable => e
+              case _ if complexGroupingExpressionMap.contains(e.canonicalized) =>
+                complexGroupingExpressionMap.get(e.canonicalized).map(_.toAttribute).getOrElse(e)
+              case _ => e.mapChildren(replaceComplexGroupingExpressions)
+            }
+          }
+
+          val newAggregateExpressions = a.aggregateExpressions
+            .map(replaceComplexGroupingExpressions(_).asInstanceOf[NamedExpression])
```

Comment:
Can it be done with a.transformExpressions similar to this one:
https://github.com/databricks/runtime/blob/8aff141a8545c2ea1759230482928e25684acbe2/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/PullOutNondeterministic.scala#L39-L41

Reply:
We need to do a manual tree traversal if we want to stop recursion earlier, e.g. `case _ if AggregateExpression.isAggregate(e) => e`

Reply:
Does the following one work?

a.transformExpressionsWithPruning(e => !(AggregateExpression.isAggregate(e) || e.fordable)) {
       ....
}

Reply:
Hmm, yes, this could work with some explicit casting.

Reply:
But this would traverse on `a.groupingExpressions` too which is not needed.

Reply:
> But this would traverse on a.groupingExpressions too which is not needed.

You're right. I think the following would behave the same as the manual recursion:

a.aggregateExpressions.map(_.transformWithPruning(e => !(AggregateExpression.isAggregate(e) || e.fordable))({
   // the first two original case branches can be skipped here.
   .....
}).asInstanceOf....)

Reply:
Anyway, it's just my small preference -- it seems neater to use framework functions if it works. Feel free to merge whatever you feel comfortable with.

Reply:
https://github.com/peter-toth/spark/commit/ed374fe046f5bb0a65b2bf72129b37df9cdce2ea could work, but I need to cast `TreePatternBits` to `Expression`.
Although, I wonder if it would make sense to split plan and expression pruning like this: https://github.com/peter-toth/spark/commit/d817fc724a78b8aa2de59999f237346d412e46dd and so this pruning became simpler: https://github.com/peter-toth/spark/commit/d817fc724a78b8aa2de59999f237346d412e46dd#diff-57201016f79912c165715811d7f7f37e2acbef2ae7b241c3c8a0b928d0052eb5R61

Reply:
Thanks a lot for exploring this, Peter!  I'll think more about such use cases.